File: 03-15 - AVL/tads/bst/BST.java
package tads.bst;

public interface BST<T extends Comparable<T>> {
  /***
   * insert inserts the data into the bst, if the data is already
   * in the bst, it does nothing
   */
  public void insert(T data);

  public void remove(T data);

  public int height();

  public int size();

  public void printInOrder();
}

File: 03-15 - AVL/tads/bst/bst.cpp

template <class T>
class BST
{
public:
  // if data is not in the tree, insert it,
  // otherwise do nothing
  virtual void insert(T data) = 0;

  virtual void remove(T data) = 0;

  virtual int height() = 0;

  virtual int size() = 0;

  virtual void printInOrder() = 0;
};
